src/k2/cmd/ycsb/data.h

```diff
@@ -0,0 +1,155 @@
+
+    The above copyright notice and this permission notice shall be included in all copies
+    or
+    substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS",
+    WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+    DAMAGES OR OTHER
+    LIABILITY,
+    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+*/
+
+#pragma once
+
+#include <decimal/decimal>
+#include <string>
+
+#include <k2/common/Common.h>
+#include <k2/module/k23si/client/k23si_client.h>
+#include <k2/transport/Payload.h>
+#include <k2/transport/PayloadSerialization.h>
+
+#include "ycsb_rand.h"
+#include "Log.h"
+using namespace k2;
+static const String ycsbCollectionName = "YCSB";
+
+// helper function to convert to base (default 93 (ascii from 33 to 126)) representation
+String itostring(int n, int base = 93)
+{
+    String buf(String::initialized_later{}, 1000);
+    std::div_t dv{}; dv.quot = n;
+    uint32_t i = 0;
+
+    do {
+        dv = std::div(dv.quot, base);
+        buf[i] = (char)(std::abs(dv.rem)+33);  // avoid ascii value 126 because it is used for padding
+        i++;
+    } while(dv.quot);
+
+    return buf.substr(0,i);
+}
+
+class YCSBData{
+
+public:
+
+    static dto::Schema generate_schema(const uint32_t num_fields){
+
+        std::vector<dto::SchemaField> field_names;
+
+        for(uint32_t field_no = 0; field_no < num_fields; ++field_no){
+            String fieldname = "field" + std::to_string(field_no);
+            dto::SchemaField f {dto::FieldType::STRING, fieldname , false, false};
+            field_names.push_back(f);
+        }
+
+        dto::Schema ycsb_data_schema {
+            .name = "ycsb_data",
+            .version = 1,
+            .fields = field_names,
+            .partitionKeyFields = std::vector<uint32_t> { 0 },
+            .rangeKeyFields = std::vector<uint32_t> {}
+        };
+
+        return ycsb_data_schema;
+    }
+
+    static String idToKey(uint64_t keyid, uint32_t len_field){
+        String key = itostring(keyid);
+
+        if(key.size()>len_field)
+            return "";
+
+        String padding( (len_field - key.size()),char(126));
+
+        return padding + key;
+    }
+
+    YCSBData(uint64_t keyid, bool generateNew = true) : ID(keyid) {
+
+        fields.push_back(YCSBData::idToKey(keyid,_len_field()));
+
+        if(fields[0]=="" || !generateNew) // key = "" means id cannot be translated to string of required length
+            return;
+
+        RandomContext random(0);
+
+        for(uint32_t field_no = 1; field_no < _num_fields(); ++field_no){
```

Comment:
Off by one here? By the way, in the TPC-C tests the only reason I have a lot of loops using 1-based indexing is because that is how the spec defines things. You can use normal 0-based indexing for YCSB

Reply:
This is because Field 0 is the key.

Reply:
I see, makes sense, thanks.